# Final demo dataset

## Loading and exploring the dataframe 

In [89]:
import pandas as pd
import os
import numpy as np

In [17]:
df = pd.read_csv("../Data/Raw/df_final_demo.csv")
df.head(20)

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gendr  num_accts  \
0      836976             6.0              73.0      60.5     U        2.0   
1     2304905             7.0              94.0      58.0     U        2.0   
2     1439522             5.0              64.0      32.0     U        2.0   
3     1562045            16.0             198.0      49.0     M        2.0   
4     5126305            12.0             145.0      33.0     F        2.0   
5     3727881             5.0              71.0      30.5     U        2.0   
6      272934             5.0              66.0      58.5     U        2.0   
7      388801            30.0             361.0      57.5     M        5.0   
8      285619            30.0             369.0      67.5     M        2.0   
9     8198645            15.0             189.0      54.5     F        2.0   
10    8368822             8.0             100.0      34.5     U        2.0   
11    2537139             8.0              99.0      51.5     U        2.0   
12    2208717             5.0              67.0      33.0     U        2.0   
13    2214073            16.0             193.0      62.5     M        3.0   
14    2368987             5.0              71.0      54.5     U        2.0   
15    2892152             4.0              58.0      36.0     U        2.0   
16    4536541             3.0              47.0      74.0     U        2.0   
17    3788589            10.0             122.0      32.5     M        2.0   
18     719792            11.0             136.0      31.0     F        2.0   
19    5904035             8.0             101.0      25.5     U        2.0   

          bal  calls_6_mnth  logons_6_mnth  
0    45105.30           6.0            9.0  
1   110860.30           6.0            9.0  
2    52467.79           6.0            9.0  
3    67454.65           3.0            6.0  
4   103671.75           0.0            3.0  
5    23915.60           0.0            3.0  
6    27021.42           2.0            5.0  
7   522498.72           1.0            4.0  
8   299388.72           3.0            6.0  
9   382303.83           6.0            9.0  
10   75578.54           0.0            3.0  
11  135940.68           6.0            9.0  
12   29782.96           6.0            9.0  
13  136818.03           6.0            9.0  
14   44408.87           2.0            5.0  
15   46815.48           5.0            8.0  
16   47804.50           1.0            4.0  
17   30750.76           0.0            3.0  
18  111736.78           5.0            8.0  
19   27848.76           3.0            6.0

In [18]:
# Checking unique values of each columns
df.nunique()


client_id           70609
clnt_tenure_yr         54
clnt_tenure_mnth      482
clnt_age              165
gendr                   4
num_accts               8
bal                 70328
calls_6_mnth            8
logons_6_mnth           9
dtype: int64

## Standardising Column names

In [19]:
# Renaming columns with typos and saving cleaned dataframe

df = df.rename(columns={
    "gendr": "gender",
    "num_acts": "num_accts",
    "bal": "clnt_balance"
})


final_demo = "../Data/Processed/df_final_demo_clean.csv"
df.to_csv("../Data/Processed/df_final_demo_clean.csv", index=False)
df = pd.read_csv(final_demo)
df.head()

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gender  num_accts  \
0     836976             6.0              73.0      60.5      U        2.0   
1    2304905             7.0              94.0      58.0      U        2.0   
2    1439522             5.0              64.0      32.0      U        2.0   
3    1562045            16.0             198.0      49.0      M        2.0   
4    5126305            12.0             145.0      33.0      F        2.0   

   clnt_balance  calls_6_mnth  logons_6_mnth  
0      45105.30           6.0            9.0  
1     110860.30           6.0            9.0  
2      52467.79           6.0            9.0  
3      67454.65           3.0            6.0  
4     103671.75           0.0            3.0

## Grouping 'gender' values

In [20]:
df["gender"].value_counts()

gender
U    24122
M    23724
F    22746
X        3
Name: count, dtype: int64

In [21]:
df[df["gender"] == "X"]


client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age gender  \
17728    4806791            11.0             136.0      40.0      X   
19475    5537155             9.0             113.0      26.0      X   
53592    1807720             9.0             110.0      52.0      X   

       num_accts  clnt_balance  calls_6_mnth  logons_6_mnth  
17728        2.0      31437.35           6.0            9.0  
19475        2.0      25380.34           1.0            4.0  
53592        3.0      25710.63           4.0            5.0

- Since we do not know the gender of these clients, we group them to "U" (Unknown).

In [22]:
df.loc[df["gender"] == "X", "gender"] = "Unknown"

In [23]:
df.loc[df["gender"] == "F", "gender"] = "Female"

In [24]:
df.loc[df["gender"] == "M", "gender"] = "Male"

In [25]:
df.loc[df["gender"] == "U", "gender"] = "Unknown"

## Dropping the null values (only the rows that have all field values as null)

In [26]:
df = df.dropna(
    subset=['clnt_tenure_yr','clnt_age','num_acts','clnt_balance','calls_6_mnth','logons_6_mnth'],
    how='all'
)

KeyError: ['num_acts']

In [ ]:
# Saving changes again
df.to_csv(final_demo, index=False)

In [ ]:
df["gender"].value_counts()

gender
Unknown    24125
Male       23724
Female     22746
Name: count, dtype: int64

In [ ]:
df.nunique()

client_id           70609
clnt_tenure_yr         54
clnt_tenure_mnth      482
clnt_age              165
gender                  3
num_acts                8
clnt_balance        70327
calls_6_mnth            8
logons_6_mnth           9
dtype: int64

### Clients Balance column inspection

* We can also see that we have 70609 client id's but only 70327 unique values for client balance, reason to be figured out 

In [ ]:
df["clnt_balance"].value_counts()


clnt_balance
31188.62     5
29388.51     4
24953.79     3
29389.11     3
29388.50     3
            ..
25561.22     1
41751.20     1
32797.00     1
375064.35    1
27021.42     1
Name: count, Length: 70327, dtype: int64

* It is because some values are shared by multiple clients

In [ ]:
# Clients who share the same balance - sanity check
df[df["clnt_balance"] == 31188.62]

client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age   gender  \
11963    1845380             4.0              58.0      26.5  Unknown   
19272    3921085             4.0              58.0      27.5  Unknown   
44906    2962021            12.0             146.0      30.0   Female   
45321    1882050             7.0              91.0      32.5  Unknown   
45538    3667266             4.0              58.0      50.0  Unknown   

       num_acts  clnt_balance  calls_6_mnth  logons_6_mnth  
11963       2.0      31188.62           1.0            4.0  
19272       2.0      31188.62           6.0            9.0  
44906       2.0      31188.62           0.0            3.0  
45321       2.0      31188.62           4.0            7.0  
45538       2.0      31188.62           2.0            5.0

In [ ]:
df[df["clnt_balance"].duplicated(keep=False)].sort_values("clnt_balance")


client_id  clnt_tenure_yr  clnt_tenure_mnth  clnt_age   gender  \
54234     940743            13.0             160.0      29.0  Unknown   
66690     928168            12.0             144.0      64.5   Female   
61571    2685682             6.0              81.0      40.5  Unknown   
54541    9035245             6.0              80.0      43.5  Unknown   
70356    5176364             4.0              57.0      29.5  Unknown   
...          ...             ...               ...       ...      ...   
46076    8611797             NaN               NaN       NaN      NaN   
47189    5144725             NaN               NaN       NaN      NaN   
49846    1037867             NaN               NaN       NaN      NaN   
63701    1618749             NaN               NaN       NaN      NaN   
66219    6021001             NaN               NaN       NaN      NaN   

       num_acts  clnt_balance  calls_6_mnth  logons_6_mnth  
54234       2.0      13885.98           1.0            1.0  
66690       2.0      13885.98           2.0            2.0  
61571       2.0      15276.78           1.0            1.0  
54541       2.0      15276.78           1.0            1.0  
70356       2.0      15735.00           2.0            2.0  
...         ...           ...           ...            ...  
46076       NaN           NaN           NaN            NaN  
47189       NaN           NaN           NaN            NaN  
49846       NaN           NaN           NaN            NaN  
63701       NaN           NaN           NaN            NaN  
66219       NaN           NaN           NaN            NaN  

[541 rows x 9 columns]

In [ ]:
df["clnt_balance"].isna().sum()


np.int64(14)

--------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------

# Final Experiment Clients Dataset

## Loading and exploring the dataframe 

In [ ]:
df_2 = pd.read_csv("../Data/Raw/df_final_experiment_clients.csv")
df_2.head(20)

client_id Variation
0     9988021      Test
1     8320017      Test
2     4033851   Control
3     1982004      Test
4     9294070   Control
5     9466803      Test
6     3170143      Test
7     6651403   Control
8     5887227      Test
9     2105948   Control
10     438567      Test
11    2519971      Test
12    9814849   Control
13    5407849      Test
14    7286739   Control
15    4446011   Control
16    5308474      Test
17    2991219      Test
18    2354414      Test
19     949210      Test

In [ ]:
df_2.isna().sum()

client_id        0
Variation    20109
dtype: int64

In [ ]:
df_2["Variation"].value_counts(dropna=False)


Variation
Test       26968
Control    23532
NaN        20109
Name: count, dtype: int64

In [ ]:
(df_2["Variation"].isna().mean() * 100).round(2)


np.float64(28.48)

* 28.5% of the DataFrame consists of NaN values

In [ ]:
df_nan = df_2[df_2["Variation"].isna()]
df_nan.head(10)

client_id Variation
50500    5459747       NaN
50501    8031000       NaN
50502    1847030       NaN
50503    9713157       NaN
50504    7775828       NaN
50505    7644910       NaN
50506    5945107       NaN
50507    2212145       NaN
50508    9755145       NaN
50509    5129355       NaN

## Replacing NaN values with "Excluded"

In [ ]:
# Changing NaN values in to "Excluded" -> This represents clients who weren't a part of the experiment.
df_2["variation"] = df_2["variation"].fillna("excluded")


KeyError: 'variation'

In [ ]:
# Changing column names to lower case and saving changes
df_2.columns = df_2.columns.str.lower()
OUTPUT_DIR = "../Data/Processed"
os.makedirs(OUTPUT_DIR, exist_ok=True)

OUTPUT_FILE = os.path.join(OUTPUT_DIR, "df_final_experiment_clients_clean.csv")
df_2.to_csv(OUTPUT_FILE, index=False)

* More than 25% of our clients weren't part of the test. This is not an error, doesn't need cleaning

-------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------

# Webdata Datasets - pt_1 and pt_2

## Loading the dataframes

In [28]:
df_3_1 = pd.read_csv("../Data/Raw/df_final_web_data_pt_1.csv")
df_3_1.head(5)

client_id            visitor_id                      visit_id process_step  \
0    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
1    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
2    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   
3    9988021  580560515_7732621733  781255054_21935453173_531117       step_2   
4    9988021  580560515_7732621733  781255054_21935453173_531117       step_3   

             date_time  
0  2017-04-17 15:27:07  
1  2017-04-17 15:26:51  
2  2017-04-17 15:19:22  
3  2017-04-17 15:19:13  
4  2017-04-17 15:18:04

In [30]:
df_3_2 = pd.read_csv("../Data/Raw/df_final_web_data_pt_2.csv")
df_3_2.head(5)

client_id             visitor_id                      visit_id  \
0     763412  601952081_10457207388  397475557_40440946728_419634   
1    6019349  442094451_91531546617  154620534_35331068705_522317   
2    6019349  442094451_91531546617  154620534_35331068705_522317   
3    6019349  442094451_91531546617  154620534_35331068705_522317   
4    6019349  442094451_91531546617  154620534_35331068705_522317   

  process_step            date_time  
0      confirm  2017-06-06 08:56:00  
1      confirm  2017-06-01 11:59:27  
2       step_3  2017-06-01 11:58:48  
3       step_2  2017-06-01 11:58:08  
4       step_1  2017-06-01 11:57:58

## Merging Final Webdata Dataframes into a single dataframe

In [32]:
# Merging 

df_3_1 = pd.read_csv("../Data/Raw/df_final_web_data_pt_1.csv")
df_3_2 = pd.read_csv("../Data/Raw/df_final_web_data_pt_2.csv")

df_final_web = pd.concat([df_3_1, df_3_2], ignore_index=True)

# Sorting by time and client 
df_final_web["date_time"] = pd.to_datetime(df_final_web["date_time"])
df_final_web = df_final_web.sort_values(["client_id", "date_time"]).reset_index(drop=True)

# Saving changes in Processed folder
OUTPUT_FILE = "../Data/Processed/df_final_web_merged.csv"
df_final_web.to_csv(OUTPUT_FILE, index=False)


In [33]:
df_final_web.head(15)

client_id             visitor_id                      visit_id  \
0         169  201385055_71273495308  749567106_99161211863_557568   
1         169  201385055_71273495308  749567106_99161211863_557568   
2         169  201385055_71273495308  749567106_99161211863_557568   
3         169  201385055_71273495308  749567106_99161211863_557568   
4         169  201385055_71273495308  749567106_99161211863_557568   
5         336    64757908_3400128256  649044751_80905125055_554468   
6         336    64757908_3400128256  649044751_80905125055_554468   
7         546  475037402_89828530214    731811517_9330176838_94847   
8         546  475037402_89828530214    731811517_9330176838_94847   
9         546  475037402_89828530214    731811517_9330176838_94847   
10        546  475037402_89828530214    731811517_9330176838_94847   
11        546  475037402_89828530214    731811517_9330176838_94847   
12        555  402506806_56087378777  637149525_38041617439_716659   
13        555  402506806_56087378777  637149525_38041617439_716659   
14        555  402506806_56087378777  637149525_38041617439_716659   

   process_step           date_time  
0         start 2017-04-12 20:19:36  
1        step_1 2017-04-12 20:19:45  
2        step_2 2017-04-12 20:20:31  
3        step_3 2017-04-12 20:22:05  
4       confirm 2017-04-12 20:23:09  
5         start 2017-06-01 07:26:55  
6         start 2017-06-01 07:42:43  
7         start 2017-06-17 10:03:29  
8        step_1 2017-06-17 10:03:39  
9        step_2 2017-06-17 10:03:52  
10       step_3 2017-06-17 10:05:19  
11      confirm 2017-06-17 10:05:42  
12        start 2017-04-15 12:57:56  
13       step_1 2017-04-15 12:58:03  
14       step_2 2017-04-15 12:58:35

## Separating Date and Time stamps

In [34]:
df_web = pd.read_csv("../Data/Processed/df_final_web_merged.csv")

df_web["date_time"] = pd.to_datetime(df_web["date_time"])

df_web["date"] = df_web["date_time"].dt.date
df_web["time"] = df_web["date_time"].dt.time

# Dropping original column
df_web = df_web.drop(columns=["date_time"])

df_web.head()

client_id             visitor_id                      visit_id  \
0        169  201385055_71273495308  749567106_99161211863_557568   
1        169  201385055_71273495308  749567106_99161211863_557568   
2        169  201385055_71273495308  749567106_99161211863_557568   
3        169  201385055_71273495308  749567106_99161211863_557568   
4        169  201385055_71273495308  749567106_99161211863_557568   

  process_step        date      time  
0        start  2017-04-12  20:19:36  
1       step_1  2017-04-12  20:19:45  
2       step_2  2017-04-12  20:20:31  
3       step_3  2017-04-12  20:22:05  
4      confirm  2017-04-12  20:23:09

In [36]:
# Saving changes

OUTPUT_DIR = "../Data/Processed"
os.makedirs(OUTPUT_DIR, exist_ok=True)

OUTPUT_FILE = os.path.join(OUTPUT_DIR, "df_final_web_merged.csv")
df_web.to_csv(OUTPUT_FILE, index=False)


## Experiment process rate count 

In [45]:
df_web.columns.tolist()


['client_id', 'visitor_id', 'visit_id', 'process_step', 'date', 'time']

In [49]:
# temporary date_time column 

df_web_2 = df_web.copy()

df_web_2["date_time"] = pd.to_datetime(
    df_web["date"].astype(str) + " " + df_web["time"].astype(str)
)


In [61]:
# Adding new cols
group_cols = ["client_id", "visitor_id", "visit_id"]


df_summary = (
    df_web_2
    .groupby(group_cols)
    .agg(
        steps_count=("process_step", "count"),
        unique_steps=("process_step", "nunique"),
        finish_time=("date_time", lambda x: x[df.loc[x.index, "process_step"] == "confirm"].max())
    )
    .reset_index()
)


In [64]:
# Removing nulls from time col and adding completion status

df_summary["completion_status"] = df_summary["finish_time"].notna()
df_summary["completion_status"] = df_summary["completion_status"].map(
    {True: "completed", False: "not_completed"}
)


In [65]:
df_completed = df_summary[df_summary["finish_time"].notna()].copy()


In [68]:
# Adding client last step column

last_steps = (
    df
    .sort_values("date_time")
    .groupby(["client_id", "visitor_id", "visit_id"])
    .last()
    .reset_index()[["client_id", "visitor_id", "visit_id", "process_step"]]
    .rename(columns={"process_step": "drop_off_step"})
)

df_summary = df_summary.merge(
    last_steps,
    on=["client_id", "visitor_id", "visit_id"],
    how="left"
)

df_summary.loc[
    df_summary["completion_status"] == "completed",
    "drop_off_step"
] = None


In [70]:
# Adding average time per step each client took in the experiment

df_web_2 = df_web_2.sort_values(
    ["client_id", "visitor_id", "visit_id", "date_time"]
)

df_web_2["step_duration_sec"] = (
    df_web_2
    .groupby(["client_id", "visitor_id", "visit_id"])["date_time"]
    .diff()
    .dt.total_seconds()
)

avg_step_time = (
    df_web_2
    .groupby(["client_id", "visitor_id", "visit_id"])["step_duration_sec"]
    .mean()
    .reset_index()
    .rename(columns={"step_duration_sec": "avg_time_per_step"})
)


In [71]:
df_summary = df_summary.merge(
    avg_step_time,
    on=["client_id", "visitor_id", "visit_id"],
    how="left"
)


In [73]:
df_summary = df_summary.drop(
    columns=["drop_off_step_x", "drop_off_step_y"]
)


In [74]:
df_summary.columns


Index(['client_id', 'visitor_id', 'visit_id', 'steps_count', 'unique_steps',
       'finish_time', 'completion_status', 'drop_off_step',
       'avg_time_per_step'],
      dtype='object')

In [75]:
OUTPUT_FILE = "../Data/Processed/df_web_visit_metrics_clean.csv"
df_summary.to_csv(OUTPUT_FILE, index=False)



In [77]:
df_summary.head(5)

client_id             visitor_id                      visit_id  \
0        169  201385055_71273495308  749567106_99161211863_557568   
1        336    64757908_3400128256  649044751_80905125055_554468   
2        546  475037402_89828530214    731811517_9330176838_94847   
3        555  402506806_56087378777  637149525_38041617439_716659   
4        647   66758770_53988066587   40369564_40101682850_311847   

   steps_count  unique_steps         finish_time completion_status  \
0            5             5 2017-04-12 20:23:09         completed   
1            2             1                 NaT     not_completed   
2            5             5 2017-06-17 10:05:42         completed   
3            5             5 2017-04-15 13:00:34         completed   
4            5             5 2017-04-12 15:47:45         completed   

  drop_off_step  avg_time_per_step  
0          None              53.25  
1           NaN             948.00  
2          None              33.25  
3          None              39.50  
4          None              94.25

* We have a few errors here like no drop off step when client did not complete the experiment and time is unreadable

In [102]:

df_web_2["date_time"] = pd.to_datetime(df_web_2["date_time"])
df_web_2 = df_web_2.sort_values(
    ["client_id", "visitor_id", "visit_id", "date_time"]
)

group_cols = ["client_id", "visitor_id", "visit_id"]


df_summary = (
    df_web_2
    .groupby(group_cols)
    .agg(
        steps_count=("process_step", "count"),
        finish_time=("date_time", lambda x: x[df_web_2.loc[x.index, "process_step"] == "confirm"].max())
    )
    .reset_index()
)

# Completion status
df_summary["completion_status"] = np.where(
    df_summary["finish_time"].notna(),
    "completed",
    "uncompleted"
)

# Client drop off step 
last_step = (
    df_web_2
    .groupby(group_cols)["process_step"]
    .last()
    .reset_index(name="last_step")
)

df_summary = df_summary.merge(
    last_step,
    on=group_cols,
    how="left",
    validate="one_to_one"
)

df_summary["drop_off_step"] = np.where(
    df_summary["completion_status"] == "completed",
    None,
    df_summary["last_step"]
)

df_summary.drop(columns="last_step", inplace=True)

# Average time per step (minutes only)

time_per_journey = (
    df_web_2
    .groupby(group_cols)["date_time"]
    .agg(lambda x: (x.max() - x.min()).total_seconds() / 60)
    .reset_index(name="total_time_minutes")
)

df_summary = df_summary.merge(
    time_per_journey,
    on=group_cols,
    how="left"
)

df_summary["avg_time_per_step"] = (
    df_summary["total_time_minutes"] / df_summary["steps_count"]
).round(2)

df_summary.drop(columns="total_time_minutes", inplace=True)

# Average time per step (minutes:seconds)
def minutes_to_mmss(x):
    if pd.isna(x):
        return None
    seconds = int(round(x * 60))
    return f"{seconds // 60}:{seconds % 60:02d}"

df_summary["avg_time_per_step_mmss"] = df_summary["avg_time_per_step"].apply(minutes_to_mmss)


df_summary["finished"] = df_summary["completion_status"] == "completed"


In [103]:
df_summary.head(15)

client_id             visitor_id                      visit_id  \
0         169  201385055_71273495308  749567106_99161211863_557568   
1         336    64757908_3400128256  649044751_80905125055_554468   
2         546  475037402_89828530214    731811517_9330176838_94847   
3         555  402506806_56087378777  637149525_38041617439_716659   
4         647   66758770_53988066587   40369564_40101682850_311847   
5         722  919259913_64837298108  984487154_55831795985_521110   
6         786  439747392_30293505155  640294664_79662229470_499394   
7         805  831412807_82548325803  451173196_11661340552_563345   
8         805  831412807_82548325803   72221519_52513154978_430287   
9         805  831412807_82548325803  905546080_75813398358_250101   
10        832  645971208_10952400740  704999523_65579948030_431837   
11        934  810392784_45004760546    7076463_57954418406_971348   
12       1028   42237450_62128060588  557292053_87239438319_391157   
13       1039  637213403_33482244512  621771101_68048173682_933803   
14       1104  194240915_18158000533  543158812_46395476577_767725   

    steps_count         finish_time completion_status drop_off_step  \
0             5 2017-04-12 20:23:09         completed          None   
1             2                 NaT       uncompleted         start   
2             5 2017-06-17 10:05:42         completed          None   
3             5 2017-04-15 13:00:34         completed          None   
4             5 2017-04-12 15:47:45         completed          None   
5             9 2017-04-19 15:00:09         completed          None   
6             6 2017-06-01 11:05:34         completed          None   
7             5                 NaT       uncompleted         start   
8             2                 NaT       uncompleted        step_1   
9             2                 NaT       uncompleted        step_1   
10            7 2017-06-16 10:40:52         completed          None   
11            4                 NaT       uncompleted         start   
12            9                 NaT       uncompleted        step_1   
13            5 2017-06-02 01:03:32         completed          None   
14            1                 NaT       uncompleted         start   

    avg_time_per_step avg_time_per_step_mmss  finished  
0                0.71                   0:43      True  
1                7.90                   7:54     False  
2                0.44                   0:26      True  
3                0.53                   0:32      True  
4                1.26                   1:16      True  
5                1.11                   1:07      True  
6                0.87                   0:52      True  
7                0.45                   0:27     False  
8                0.02                   0:01     False  
9                0.67                   0:40     False  
10               0.32                   0:19      True  
11               0.59                   0:35     False  
12               1.00                   1:00     False  
13               0.69                   0:41      True  
14               0.00                   0:00     False